src/rebar3_hex_app.erl

```diff
@@ -0,0 +1,205 @@
+    lists:foldl(fun(D, Acc) -> lock_to_dep(D, Acc) end, [], Deps).
+
+lock_to_dep({A, {pkg, N, V, _, _}, 0}, Acc) ->
+    [{N, [{<<"app">>, A}, {<<"optional">>, false}, {<<"requirement">>, V}]} | Acc];
+lock_to_dep({A, {pkg, N, V, _}, 0}, Acc) ->
+    [{N, [{<<"app">>, A}, {<<"optional">>, false}, {<<"requirement">>, V}]} | Acc];
+lock_to_dep(_, Acc) ->
+    Acc.
+
+-spec validate(app_data()) -> ok | {error, validation_errors()}.
+validate(AppData) ->
+    F = fun(K, Acc) ->
+        case validate_app(K, AppData) of
+            ok ->
+                Acc;
+            {warn, Warn} ->
+                #{warnings := WarnList} = Acc,
+                maps:put(warnings, [Warn | WarnList], Acc);
+            {error, Error} ->
+                #{errors := ErrList} = Acc,
+                maps:put(errors, [Error | ErrList], Acc)
+        end
+    end,
+    case lists:foldl(F, #{warnings => [], errors => []}, ?VALIDATIONS) of
+        [] ->
+            ok;
+        Errors ->
+            {error, Errors}
+    end.
+
+validate_app(has_unstable_deps, #{deps := Deps}) ->
+    case lists:foldl(fun(Dep, Acc) -> is_unstable_dep(Dep, Acc) end, [], Deps) of
+        [] ->
+            ok;
+        PreDeps ->
+            {warn, {has_unstable_deps, PreDeps}}
+    end;
+validate_app(has_semver, #{name := Name, version := Ver}) ->
+    case rebar3_hex_version:parse(Ver) of
+        {error, invalid_version} ->
+            {error, {invalid_semver, {Name, Ver}}};
+        _ ->
+            ok
+    end;
+validate_app(has_contributors, #{name := Name, details := AppDetails}) ->
+    case proplists:is_defined(contributors, AppDetails) of
+        true ->
+            {warn, {has_contributors, Name}};
+        false ->
+            ok
+    end;
+validate_app(has_maintainers, #{name := Name, details := AppDetails}) ->
+    case proplists:is_defined(maintainers, AppDetails) of
+        true ->
+            {warn, {has_maintainers, Name}};
+        false ->
+            ok
+    end;
+validate_app(has_description, #{name := Name, details := AppDetails}) ->
+    case is_empty_prop(description, AppDetails) of
+        true ->
+            {warn, {no_description, Name}};
+        false ->
+            ok
+    end;
+validate_app(has_licenses, #{name := Name, details := AppDetails}) ->
+    case is_empty_prop(licenses, AppDetails) of
+        true ->
+            {warn, {no_license, Name}};
+        _ ->
+            ok
+    end.
+
+is_empty_prop(K, PropList) ->
+    Prop = proplists:get_value(K, PropList),
+    case Prop of
+        Empty when Empty =:= [] orelse Empty =:= undefined ->
+            true;
+        _ ->
+            false
+    end.
+
+is_unstable_dep({_, {pkg, Pkg, Ver, _, _}, _}, Acc) ->
+    case rebar3_hex_version:parse(Ver) of
+        {ok, #{pre := Pre}} when Pre =/= [] ->
+            [{Pkg, Ver} | Acc];
+        _ ->
+            Acc
+    end;
+is_unstable_dep(_, Acc) ->
+    Acc.
+
+-spec vcs_vsn(rebar_state:t(), rebar_app_info:t()) -> string().
```

Comment:
This function comes from `rebar_utils`.  It is deprecated and was only kept in rebar3 for rebar3_hex, just went ahead and moved it on in so that rebar3 may be able to remove it now.